File: autograd.go
package main

import "math"

// Value is a node in the autograd computation graph.
type Value struct {
	Data     float64
	Grad     float64
	m        float64 // momentum / Adam first moment
	v        float64 // Adam second moment
	prev     []*Value
	backward func()
}

func newValue(data float64) *Value {
	return &Value{Data: data, backward: func() {}}
}

func mulV(a, b *Value) *Value {
	out := &Value{Data: a.Data * b.Data, prev: []*Value{a, b}}
	out.backward = func() {
		a.Grad += b.Data * out.Grad
		b.Grad += a.Data * out.Grad
	}
	return out
}

func addV(a, b *Value) *Value {
	out := &Value{Data: a.Data + b.Data, prev: []*Value{a, b}}
	out.backward = func() {
		a.Grad += out.Grad
		b.Grad += out.Grad
	}
	return out
}

func subV(a, b *Value) *Value {
	out := &Value{Data: a.Data - b.Data, prev: []*Value{a, b}}
	out.backward = func() {
		a.Grad += out.Grad
		b.Grad -= out.Grad
	}
	return out
}

func powV(a *Value, n float64) *Value {
	out := &Value{Data: math.Pow(a.Data, n), prev: []*Value{a}}
	out.backward = func() {
		a.Grad += out.Grad * n * math.Pow(a.Data, n-1)
	}
	return out
}

func tanhV(a *Value) *Value {
	t := math.Tanh(a.Data)
	out := &Value{Data: t, prev: []*Value{a}}
	out.backward = func() {
		th := math.Tanh(a.Data)
		a.Grad += out.Grad * (1 - th*th)
	}
	return out
}

func buildTopo(v *Value, visited map[*Value]bool, topo *[]*Value) {
	if !visited[v] {
		visited[v] = true
		for _, child := range v.prev {
			buildTopo(child, visited, topo)
		}
		*topo = append(*topo, v)
	}
}

func (v *Value) doBackward() {
	topo := make([]*Value, 0, 64)
	visited := make(map[*Value]bool)
	buildTopo(v, visited, &topo)
	v.Grad = 1.0
	for i := len(topo) - 1; i >= 0; i-- {
		topo[i].backward()
	}
}

func (v *Value) step(lr float64) {
	topo := make([]*Value, 0, 64)
	visited := make(map[*Value]bool)
	buildTopo(v, visited, &topo)
	noop := func() {}
	for _, node := range topo {
		if len(node.prev) == 0 {
			node.Data -= lr * node.Grad
			node.Grad = 0.0
		} else {
			node.backward = noop
			node.prev = nil
		}
	}
}

func (v *Value) momentumStep(lr, beta float64) {
	topo := make([]*Value, 0, 64)
	visited := make(map[*Value]bool)
	buildTopo(v, visited, &topo)
	noop := func() {}
	for _, node := range topo {
		if len(node.prev) == 0 {
			node.m = beta*node.m + node.Grad
			node.Data -= lr * node.m
			node.Grad = 0.0
		} else {
			node.backward = noop
			node.prev = nil
		}
	}
}

func (v *Value) adamStep(lr float64, t int, beta1, beta2, eps float64) {
	topo := make([]*Value, 0, 64)
	visited := make(map[*Value]bool)
	buildTopo(v, visited, &topo)
	noop := func() {}
	for _, node := range topo {
		if len(node.prev) == 0 {
			node.m = beta1*node.m + (1-beta1)*node.Grad
			node.v = beta2*node.v + (1-beta2)*node.Grad*node.Grad
			mHat := node.m / (1 - math.Pow(beta1, float64(t)))
			vHat := node.v / (1 - math.Pow(beta2, float64(t)))
			node.Data -= lr * mHat / (math.Sqrt(vHat) + eps)
			node.Grad = 0.0
		} else {
			node.backward = noop
			node.prev = nil
		}
	}
}

File: autograd.py
import math

class Value:
    def __init__(self, data, _prev=[]):
        self.data = data
        self.grad = 0.0
        self._prev = _prev
        self._backward = lambda: None

    def __repr__(self):
        return f"V({self.data},{self.grad})"

    @staticmethod
    def mul(a, b):
        out = Value(a.data * b.data, _prev=[a, b])
        def _backward():
            a.grad += b.data * out.grad
            b.grad += a.data * out.grad
        out._backward = _backward
        return out

    @staticmethod
    def add(a, b):
        out = Value(a.data + b.data, _prev=[a, b])
        def _backward():
            a.grad += out.grad
            b.grad += out.grad
        out._backward = _backward
        return out

    @staticmethod
    def sub(a, b):
        out = Value(a.data - b.data, _prev=[a, b])
        def _backward():
            a.grad += out.grad
            b.grad -= out.grad
        out._backward = _backward
        return out

    @staticmethod
    def pow(a, n):
        out = Value(pow(a.data, n), _prev=[a])
        def _backward():
            a.grad += out.grad*n*pow(a.data, n-1)
        out._backward = _backward
        return out

    @staticmethod
    def relu(a):
        out = Value(max(0.0, a.data), _prev=[a])
        def _backward():
            a.grad += out.grad * (1.0 if a.data > 0 else 0.0)
        out._backward = _backward
        return out

    @staticmethod
    def elu(a, alpha=1.0):
        v = a.data if a.data > 0 else alpha * (math.exp(a.data) - 1)
        out = Value(v, _prev=[a])
        def _backward():
            a.grad += out.grad * (1.0 if a.data > 0 else alpha * math.exp(a.data))
        out._backward = _backward
        return out

    @staticmethod
    def silu(a):
        sig = 1.0 / (1.0 + math.exp(-a.data))
        out = Value(a.data * sig, _prev=[a])
        def _backward():
            sig = 1.0 / (1.0 + math.exp(-a.data))
            a.grad += out.grad * (sig + a.data * sig * (1.0 - sig))
        out._backward = _backward
        return out

    @staticmethod
    def sin(a):
        out = Value(math.sin(a.data), _prev=[a])
        def _backward():
            a.grad += out.grad * math.cos(a.data)
        out._backward = _backward
        return out

    @staticmethod
    def tanh(a):
        out = Value(math.tanh(a.data), _prev=[a])
        def _backward():
            a.grad += out.grad*(1 - math.tanh(a.data)**2)
        out._backward = _backward
        return out

    def backward(self):
        topo = []
        visited = set()
        def build(v):
            if v not in visited:
                visited.add(v)
                for child in v._prev:
                    build(child)
                topo.append(v)
        build(self)
        self.grad = 1.0
        for node in reversed(topo):
            node._backward()

    def step(self, lr):
        topo = []
        visited = set()
        def build(v):
            if v not in visited:
                visited.add(v)
                for child in v._prev:
                    build(child)
                topo.append(v)
        build(self)
        for node in topo:
            if not node._prev:
                node.data -= lr * node.grad
                node.grad = 0.0
            else:
                node._backward = lambda: None
                node._prev = []
